subsys/bluetooth/services/ots_oacp.c

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/printk.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
+
+#include <bluetooth/gatt.h>
+#include <bluetooth/services/ots.h>
+#include "ots_internal.h"
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(bt_gatt_ots, CONFIG_BT_GATT_OTS_LOG_LEVEL);
+
+#define OACP_PROC_TYPE_SIZE	1
+#define OACP_RES_MAX_SIZE	3
+
+static enum bt_gatt_ots_oacp_res_code oacp_read_proc_validate(
+	struct bt_conn *conn,
+	struct bt_gatt_ots *ots,
+	struct bt_gatt_ots_oacp_proc *proc)
+{
+	struct bt_gatt_ots_oacp_read_params *params = &proc->read_params;
+
+	LOG_DBG("Validating Read procedure with offset: 0x%08X and "
+		"length: 0x%08X", params->offset, params->len);
+
+	if (!ots->cur_obj) {
+		return BT_GATT_OTS_OACP_RES_INV_OBJ;
+	}
+
+	if (!BT_GATT_OTS_OBJ_GET_PROP_READ(ots->cur_obj->metadata.props)) {
+		return BT_GATT_OTS_OACP_RES_NOT_PERMITTED;
+	}
+
+	if (!bt_gatt_ots_l2cap_is_open(&ots->l2cap, conn)) {
+		return BT_GATT_OTS_OACP_RES_CHAN_UNAVAIL;
+	}
+
+	if ((params->offset + (uint64_t) params->len) >
+		ots->cur_obj->metadata.size.cur) {
+		return BT_GATT_OTS_OACP_RES_INV_PARAM;
+	}
+
+	if (ots->cur_obj->state.type != BT_GATT_OTS_OBJECT_IDLE_STATE) {
+		return BT_GATT_OTS_OACP_RES_OBJ_LOCKED;
+	}
+
+	ots->cur_obj->state.type = BT_GATT_OTS_OBJECT_READ_OP_STATE;
+	ots->cur_obj->state.read_op.sent_len = 0;
+	memcpy(&ots->cur_obj->state.read_op.oacp_params, &proc->read_params,
+		sizeof(ots->cur_obj->state.read_op.oacp_params));
+
+	LOG_DBG("Read procedure is accepted");
+
+	return BT_GATT_OTS_OACP_RES_SUCCESS;
+}
+
+static enum bt_gatt_ots_oacp_res_code oacp_proc_validate(
+	struct bt_conn *conn,
+	struct bt_gatt_ots *ots,
+	struct bt_gatt_ots_oacp_proc *proc)
+{
+	switch (proc->type) {
+	case BT_GATT_OTS_OACP_PROC_READ:
+		return oacp_read_proc_validate(conn, ots, proc);
+	case BT_GATT_OTS_OACP_PROC_CREATE:
+	case BT_GATT_OTS_OACP_PROC_DELETE:
+	case BT_GATT_OTS_OACP_PROC_CHECKSUM_CALC:
+	case BT_GATT_OTS_OACP_PROC_EXECUTE:
+	case BT_GATT_OTS_OACP_PROC_WRITE:
+	case BT_GATT_OTS_OACP_PROC_ABORT:
+	default:
+		return BT_GATT_OTS_OACP_RES_OPCODE_NOT_SUP;
+	}
+};
+
+static enum bt_gatt_ots_oacp_res_code oacp_command_decode(
+	const uint8_t *buf, uint16_t len, struct bt_gatt_ots_oacp_proc *proc)
+{
+	const uint8_t *end_buf = (buf + len);
+
+	proc->type = *buf++;
```

Comment:
Consider using the `net_buf_simple` struct and functions to pull the data, rather than manually calling `sys_get_*` and `buf += ...`

Reply:
I think `net_buf_simple_init_with_data` would be useful here.